encodings/alp/src/alp.rs

```diff
@@ -103,56 +103,126 @@ pub trait ALPFloat: Float + Display + 'static {
-            let decoded = Self::decode_single(e, exponents);
-            if decoded == value {
-                return Ok(e);
-            }
+        let encoded = unsafe { Self::encode_single_unchecked(value, exponents) };
+        let decoded = Self::decode_single(encoded, exponents);
+        if decoded == value {
+            return Ok(encoded);
         }
-
         Err(value)
     }
 
     #[inline]
     fn decode_single(encoded: Self::ALPInt, exponents: Exponents) -> Self {
-        let encoded_float: Self = Self::from(encoded).unwrap_or_else(|| {
-            vortex_panic!(
-                "Failed to convert encoded value {} from {} to {} in ALPFloat::decode_single",
-                encoded,
-                std::any::type_name::<Self::ALPInt>(),
-                std::any::type_name::<Self>()
-            )
-        });
-        encoded_float * Self::F10[exponents.f as usize] * Self::IF10[exponents.e as usize]
+        Self::from_int(encoded) * Self::F10[exponents.f as usize] * Self::IF10[exponents.e as usize]
+    }
+
+    /// # Safety
+    ///
+    /// The returned value may not decode back to the original value.
+    #[inline(always)]
+    unsafe fn encode_single_unchecked(value: Self, exponents: Exponents) -> Self::ALPInt {
+        (value * Self::F10[exponents.e as usize] * Self::IF10[exponents.f as usize])
+            .fast_round()
+            .as_int()
+    }
+}
+
+fn encode_chunk_unchecked<T: ALPFloat>(
+    chunk: &[T],
+    exp: Exponents,
+    encoded_output: &mut Vec<T::ALPInt>,
+    patch_indices: &mut Vec<u64>,
+    patch_values: &mut Vec<T>,
+    fill_value: &mut Option<T::ALPInt>,
+) {
+    let num_prev_encoded = encoded_output.len();
+    let num_prev_patches = patch_indices.len();
+    assert_eq!(patch_indices.len(), patch_values.len());
+    let has_filled = fill_value.is_some();
+
+    // encode the chunk, counting the number of patches
+    let mut chunk_patch_count = 0;
+    encoded_output.extend(chunk.iter().map(|v| {
+        let encoded = unsafe { T::encode_single_unchecked(*v, exp) };
+        let decoded = T::decode_single(encoded, exp);
+        let neq = (decoded != *v) as usize;
+        chunk_patch_count += neq;
+        encoded
+    }));
+    let chunk_patch_count = chunk_patch_count; // immutable hereafter
+    assert_eq!(encoded_output.len(), num_prev_encoded + chunk.len());
+
+    // find the first successfully encoded value (i.e., not patched)
+    // this is our fill value for missing values
+    if fill_value.is_none() && (num_prev_encoded + chunk_patch_count < encoded_output.len()) {
+        assert_eq!(num_prev_encoded, num_prev_patches);
+        for i in num_prev_encoded..encoded_output.len() {
+            if i >= patch_indices.len() || patch_indices[i] != i as u64 {
+                *fill_value = Some(encoded_output[i]);
+                break;
+            }
+        }
+    }
+
+    // if there are no patches, we are done
+    if chunk_patch_count == 0 {
```

Comment:
Need to handle the edge case of 2 chunks where chunk 0 is all patches, chunk 1 has 0 patches... which won't fill